procedures: do not reference Che-Theia in Importing untrusted TLS certificates (#2507)
codenvy/che-docs: modules/administration-guide/pages/importing-untrusted-tls-certificates.adoc

Therefore, you must import into {prod-short} all untrusted CA chains in use by an external service such as:

* A proxy
* An identity provider (OIDC)
* A source code repositories provider (Git)

{prod-short} uses labeled config maps in {prod-short} {orch-namespace} as sources for TLS certificates.
The config maps can have an arbitrary amount of keys with a random amount of certificates each.

[NOTE]
====
When an OpenShift cluster contains cluster-wide trusted CA certificates added through the link:https://docs.openshift.com/container-platform/latest/networking/configuring-a-custom-pki.html#nw-proxy-configure-object_configuring-a-custom-pki[cluster-wide-proxy configuration],
{prod-short} Operator detects them and automatically injects them into a config map with the `config.openshift.io/inject-trusted-cabundle="true"` label.
Based on this annotation, OpenShift automatically injects the cluster-wide trusted CA certificates inside the `ca-bundle.crt` key of the config map.
====

.Prerequisites
* An active `{orch-cli}` session with administrative permissions to the destination {orch-name} cluster.
See {orch-cli-link}.

* The `{prod-namespace}` {orch-namespace} exists.

> * For each CA chain to import, the root CA and intermediate certificates, in link:https://wiki.openssl.org/index.php/PEM[PEM] format, in a `ca-cert-for-{prod-id-short}-__<count>__.pem` file.

Author: ```suggestion
* For each CA chain to import: the root CA and intermediate certificates, in link:https://wiki.openssl.org/index.php/PEM[PEM] format, in a `ca-cert-for-{prod-id-short}-__<count>__.pem` file.
```

[id="importing-untrusted-tls-certificates"]
= Importing untrusted TLS certificates to {prod-short}

{prod-short} components communications with following services are encrypted with TLS.
They require TLS certificates signed by trusted Certificate Authorities (CA):

> * External communications between {prod-short} components

Reviewer: I don't think it's possible to configure Che to use "external inter-component communications" anymore (every component communication is done using services, not routes or ingresses). @tolusha can keep me honest here.

Author: If that's the case, great, the explanation will be much simpler then.

Reviewer 2: There is no external communication between Che components anymore

Author: Thanks, modified.

* Communications of {prod-short} components with external services

Therefore, you must import the full Certificate Authorities chain (root CA and intermediate) into the {prod-short} instance when the certificates used by one of such services are signed by an otherwise untrusted CA:

* The underlying {orch-name} cluster
* An external proxy
* An external identity provider (OIDC)
* An external source code repositories provider (Git)

> {prod-short} uses labeled config maps in {prod-short} {orch-namespace} as sources for TLS certificates.

Reviewer: I would s/config maps/ConfigMaps/g (c.f. https://kubernetes.io/docs/concepts/configuration/configmap/)

Author: Unfortunately, Kuberbenetes and OpenShift writing convention diverge. Downstream we have to comply with: https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/term_glossary.adoc#config-map and https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#api-object-formatting

> a config map

or

> the `ConfigMap` object, the `ConfigMap` resource, 

Maybe we could also use an attribute to apply different conventions upstream and downstream?

The config maps can have an arbitrary amount of keys with a random amount of certificates each.

[NOTE]
====
When an OpenShift cluster contains cluster-wide trusted CA certificates added through the link:https://docs.openshift.com/container-platform/latest/networking/configuring-a-custom-pki.html#nw-proxy-configure-object_configuring-a-custom-pki[cluster-wide-proxy configuration],
{prod-short} Operator detects them and automatically injects them into a config map with the `config.openshift.io/inject-trusted-cabundle="true"` label.
Based on this annotation, OpenShift automatically injects the cluster-wide trusted CA certificates inside the `ca-bundle.crt` key of the config map.
====

.Prerequisites
* An active `{orch-cli}` session with administrative permissions to the destination {orch-name} cluster.
See {orch-cli-link}.

* The `{prod-namespace}` {orch-namespace} exists.

> * The complete Certificate Authorities chains to import (root CA and intermediate) in link:https://wiki.openssl.org/index.php/PEM[PEM] format in `ca-cert-__<count>__.pem` files.

Reviewer 3: ```suggestion
* The complete Certificate Authorities chains (root CA and intermediate) are ready for import in link:https://wiki.openssl.org/index.php/PEM[PEM] format in `ca-cert-__<count>__.pem` files.
```

Author: I believe with the suggestion we loose the meaning: we must have this: "the complete Certificate Authorities chains to import", with additional conditions:  "in PEM format", and "in ca-certs-<coun>.pem files".

Reviewer 3: ok, I see. Leave it as is, then.

Author: I tried something for clarity.

See {orch-cli-link}.

* An active `{orch-cli}` session with administrative permissions to the destination {orch-name} cluster. See {orch-cli-link}.
* Namespace for {prod-short} exists.
* {prod-short} already uses some reserved file names to automatically inject certificates into the `ConfigMap` object. Avoid using the following reserved file names to save your certificates:
  ** `ca-bundle.crt`
  ** `ca.crt`
* The `{prod-namespace}` {orch-namespace} exists.

.Procedure
* The complete Certificate Authorities chains to import (root CA and intermediate) in link:https://wiki.openssl.org/index.php/PEM[PEM] format in `ca-cert-__<count>__.pem` files.

. Save the certificates you need to import to a local file system.
.Procedure
. Concatenate all CA chains PEM files to import, into the `custom-ca-certificates.pem` file, and remove the return character that is incompatible with the Java trust store.
+
[CAUTION]
====
* A certificate with the introductory phrase `BEGIN TRUSTED CERTIFICATE` is likely in the PEM `TRUSTED CERTIFICATE` format, which is not supported by Java. Convert it to the supported `CERTIFICATE` format with the following command:
** `openssl x509 -in cert.pem -out cert.cer`

> ====

Reviewer: Replacing carriage return chars addresses the `TRUSTED CERTIFICATE` format not being supported by Java? That would explain why you are removing this "CAUTION" admonition. Otherwise the admonition is still useful in my opinion.

Author: Yes. Typically, users will download the CA chain from a browser such as Firefox, and the browser adds the carriage return. Then the certificates are ignored by Java. (I tested with the same CA chain, with and without the carriage return).

The mention about Base64 encoding was a leftover from a previous version of the docs: when you apply a YAML file (rather than referencing a PEM file in the CLI), you must re-encode the PEM certificate in Base64 in the YAML. Applying a YAML file is better for idempotency. But the method proposed now is better for clarity.

Author: Additionally, the previous paragraph was:
* pointless, as the command output is a *PEM* certificate. See openssl x509 man page.
* destructive, as the command output is *one* certificate (leaving out the intermediates if any).

.Verification steps
. Verify that the config map contains your custom CA certificates.

> The command returns your custom CA certificates in PEM format:

Reviewer 3: ```suggestion
This command returns your custom CA certificates in PEM format:
```

+
[subs="+attributes,+quotes",options="nowrap",role=white-space-pre]
----
$ {orch-cli} get configmap \
    --namespace={prod-namespace} \
    --output='jsonpath={.items[0:].data.custom-ca-certificates\.pem}' \
    --selector=app.kubernetes.io/component=ca-bundle,app.kubernetes.io/part-of=che.eclipse.org
----

. Verify {prod-short} pod contains a volume mounting the `ca-certs-merged` config map:
+
[subs="+attributes,+quotes",options="nowrap",role=white-space-pre]
----
$ {orch-cli} get pod \
    --selector=app.kubernetes.io/component={prod-id-short} \
    --output='jsonpath={.items[0].spec.volumes[0:].configMap.name}' \
    --namespace={prod-namespace} \
    | grep ca-certs-merged
----

. Verify the {prod-short} server container has your custom CA certificates.

> The command returns your custom CA certificates in PEM format:

Reviewer 3: ```suggestion
This command returns your custom CA certificates in PEM format:
```

+
[subs="+attributes,+quotes",options="nowrap",role=white-space-pre]
----
$ {orch-cli} exec -t deploy/{prod-id-short} \
    --namespace={prod-namespace} \
    -- cat /public-certs/custom-ca-certificates.pem
----

. Verify in the {prod-short} server logs that the imported certificates count is not null:
+
[subs="+attributes,+quotes",options="nowrap",role=white-space-pre]
----
$ {orch-cli} logs deploy/{prod-id-short} --namespace={prod-namespace} \
    | grep custom-ca-certificates.pem
----

. List the SHA256 fingerprints of your certificates:
+
[subs="+attributes,+quotes",options="nowrap",role=white-space-pre]
----
$ for certificate in ca-cert*.pem ;
  do openssl x509 -in $certificate -digest -sha256 -fingerprint -noout | cut -d= -f2;
  done
----

> . Verify {prod-short} server Java truststore contains certificates with the same fingerprint:

Reviewer 3: ```suggestion
. Verify  that {prod-short} server Java truststore contains certificates with the same fingerprint:
```

+
[subs="+attributes,+quotes",options="nowrap",role=white-space-pre]
----
$ {orch-cli} exec -t deploy/{prod-id-short} --namespace={prod-namespace} -- \
    keytool -list -keystore {prod-home}/cacerts \
    | grep --after-context=1 custom-ca-certificates.pem
----

. Start a workspace, get the {orch-namespace} name in which it has been created: __<workspace_namespace>__, and wait for the workspace to be started.

. Verify that the `che-trusted-ca-certs` config map contains your custom CA certificates.

> The command returns your custom CA certificates in PEM format:

Reviewer 3: ```suggestion
This command returns your custom CA certificates in PEM format:
```

+
[subs="+attributes,+quotes",options="nowrap",role=white-space-pre]
----
$ {orch-cli} get configmap che-trusted-ca-certs \
    --namespace=__<workspace_namespace>__ \
    --output='jsonpath={.data.custom-ca-certificates\.custom-ca-certificates\.pem}'
----

. Verify that the workspace pod mounts the `che-trusted-ca-certs` config map:
+
[subs="+attributes,+quotes",options="nowrap",role=white-space-pre]
----
$ {orch-cli} get pod \
    --namespace=__<workspace_namespace>__ \
    --selector='controller.devfile.io/devworkspace_name=__<workspace_name>__' \
    --output='jsonpath={.items[0:].spec.volumes[0:].configMap.name}' \
    | grep che-trusted-ca-certs
----

. Verify that the `universal-developer-image` container (or the container defined in the workspace devfile) mounts the `che-trusted-ca-certs` volume:
+
[subs="+attributes,+quotes",options="nowrap",role=white-space-pre]
----
$ {orch-cli} get pod \
    --namespace=__<workspace_namespace>__ \
    --selector='controller.devfile.io/devworkspace_name=__<workspace_name>__' \
    --output='jsonpath={.items[0:].spec.containers[0:]}' \
    | jq 'select (.volumeMounts[].name == "che-trusted-ca-certs") | .name'
----

> . Get __<workspace_pod_name>__, the workspace pod name:

Reviewer: I would retrieving the workspace pod name earlier. Before running the commands to look for pod volumes and container volumemounts. The commands would be simplified.

Author: That was my intention at the beginning, but it's the only command that doesn't have a variant that uses only values that the Che Dashboard displays. 

I believe it makes sense to use these values as much as possible.

I would prefer to find such a variant for this last step.

> . Get __<workspace_pod_name>__, the workspace pod name:

Reviewer 3: ```suggestion
. Get the workspace pod name __<workspace_pod_name>__:
```

+
[subs="+attributes,+quotes",options="nowrap",role=white-space-pre]
----
$ {orch-cli} get pod \
    --namespace=__<workspace_namespace>__ \
    --selector='controller.devfile.io/devworkspace_name=__<workspace_name>__' \
    --output='jsonpath={.items[0:].metadata.name}' \
----

> . Verify the workspace container has your custom CA certificates.

Reviewer 3: ```suggestion
. Verify that the workspace container has your custom CA certificates.
```

> The command returns your custom CA certificates in PEM format:

Reviewer 3: ```suggestion
This command returns your custom CA certificates in PEM format:
```
